File: packages/face-detection/android/src/main/java/io/capawesome/capacitorjs/plugins/mlkit/facedetection/classes/ProcessImageResult.java
package io.capawesome.capacitorjs.plugins.mlkit.facedetection.classes;

import android.graphics.PointF;
import android.graphics.Rect;
import com.getcapacitor.JSArray;
import com.getcapacitor.JSObject;
import com.google.mlkit.vision.face.Face;
import com.google.mlkit.vision.face.FaceContour;
import com.google.mlkit.vision.face.FaceLandmark;
import java.util.List;

public class ProcessImageResult {

    private List<Face> faces;

    public ProcessImageResult(List<Face> faces) {
        this.faces = faces;
    }

    public JSObject toJSObject() {
        JSArray facesResult = this.createFacesResult();

        JSObject result = new JSObject();
        result.put("faces", facesResult);
        return result;
    }

    private JSArray createFacesResult() {
        JSArray result = new JSArray();
        for (Face face : faces) {
            JSObject faceResult = this.createFaceResult(face);
            result.put(faceResult);
        }
        return result;
    }

    private JSObject createFaceResult(Face face) {
        JSObject result = new JSObject();

        Rect boundingBox = face.getBoundingBox();
        JSObject boundsResult = this.createBoundsResult(boundingBox);
        result.put("bounds", boundsResult);

        JSArray landmarksResult = new JSArray();
        for (FaceLandmark landmark : face.getAllLandmarks()) {
            JSObject landmarkResult = this.createLandmarkResult(landmark);
            landmarksResult.put(landmarkResult);
        }
        if (landmarksResult.length() > 0) {
            result.put("landmarks", landmarksResult);
        }

        JSArray contoursResult = new JSArray();
        for (FaceContour contour : face.getAllContours()) {
            List<PointF> points = contour.getPoints();
            JSArray pointsResult = new JSArray();
            for (PointF point : points) {
                JSObject positionResult = this.createPositionResult(point);
                pointsResult.put(positionResult);
            }
            if (pointsResult.length() > 0) {
                JSObject contourResult = this.createContourResult(contour, pointsResult);
                contoursResult.put(contourResult);
            }
        }
        if (contoursResult.length() > 0) {
            result.put("contours", contoursResult);
        }

        Integer trackingId = face.getTrackingId();
        if (trackingId != null) {
            result.put("trackingId", trackingId);
        }

        result.put("headEulerAngleX", face.getHeadEulerAngleX());
        result.put("headEulerAngleY", face.getHeadEulerAngleY());
        result.put("headEulerAngleZ", face.getHeadEulerAngleZ());

        Float smilingProbability = face.getSmilingProbability();
        if (smilingProbability != null) {
            result.put("smilingProbability", smilingProbability);
        }

        Float leftEyeOpenProbability = face.getLeftEyeOpenProbability();
        if (leftEyeOpenProbability != null) {
            result.put("leftEyeOpenProbability", leftEyeOpenProbability);
        }

        Float rightEyeOpenProbability = face.getRightEyeOpenProbability();
        if (rightEyeOpenProbability != null) {
            result.put("rightEyeOpenProbability", rightEyeOpenProbability);
        }

        return result;
    }

    private JSObject createBoundsResult(Rect boundingBox) {
        JSObject result = new JSObject();
        result.put("left", boundingBox.left);
        result.put("top", boundingBox.top);
        result.put("right", boundingBox.right);
        result.put("bottom", boundingBox.bottom);
        return result;
    }

    private JSObject createLandmarkResult(FaceLandmark landmark) {
        JSObject positionResult = this.createPositionResult(landmark.getPosition());
        JSObject result = new JSObject();
        result.put("type", landmark.getLandmarkType());
        result.put("position", positionResult);
        return result;
    }

    private JSObject createPositionResult(PointF point) {
        JSObject result = new JSObject();
        result.put("x", point.x);
        result.put("y", point.y);
        return result;
    }

    private JSObject createContourResult(FaceContour contour, JSArray pointsResult) {
        JSObject result = new JSObject();
        result.put("type", contour.getFaceContourType());
        result.put("points", pointsResult);
        return result;
    }
}

File: packages/face-detection/ios/Plugin/Classes/ProcessImageResult.swift
import Foundation
import Capacitor
import MLKitVision
import MLKitFaceDetection

@objc class ProcessImageResult: NSObject {
    let faces: [Face]

    init(faces: [Face]) {
        self.faces = faces
    }

    func toJSObject() -> JSObject {
        let facesResult = createFacesResult()

        var result = JSObject()
        result["faces"] = facesResult

        return result
    }

    private func createFacesResult() -> JSArray {
        var result = JSArray()

        for face in faces {
            let faceResult = createFaceResult(face)
            result.append(faceResult)
        }

        return result
    }

    private func createFaceResult(_ face: Face) -> JSObject {
        var result = JSObject()

        let frame = face.frame
        let boundsResult = createBoundsResult(frame)
        result["bounds"] = boundsResult

        var landmarksResult = [[String: Any]]()
        for landmark in face.landmarks {
            let landmarkResult = createLandmarkResult(landmark)
            landmarksResult.append(landmarkResult)
        }
        if landmarksResult.count > 0 {
            result["landmarks"] = landmarksResult
        }

        var contoursResult = JSArray()
        for contour in face.contours {
            let pointsResult = createPointsResult(contour.points)
            if pointsResult.count > 0 {
                let contourResult = createContourResult(contour, pointsResult)
                contoursResult.append(contourResult)
            }
        }
        if contoursResult.count > 0 {
            result["contours"] = contoursResult
        }

        if face.hasTrackingID {
            result["trackingId"] = face.trackingID
        }

        if face.hasHeadEulerAngleX {
            result["headEulerAngleX"] = Float(face.headEulerAngleX)
        }

        if face.hasHeadEulerAngleY {
            result["headEulerAngleY"] = Float(face.headEulerAngleY)
        }

        if face.hasHeadEulerAngleZ {
            result["headEulerAngleZ"] = Float(face.headEulerAngleZ)
        }

        if face.hasSmilingProbability {
            result["smilingProbability"] = Float(face.smilingProbability)
        }

        if face.hasLeftEyeOpenProbability {
            result["leftEyeOpenProbability"] = Float(face.leftEyeOpenProbability)
        }

        if face.hasRightEyeOpenProbability {
            result["rightEyeOpenProbability"] = Float(face.rightEyeOpenProbability)
        }

        return result
    }

    private func createBoundsResult(_ frame: CGRect) -> JSObject {
        var result = JSObject()
        result["left"] = Float(frame.origin.x)
        result["top"] = Float(frame.origin.y)
        result["right"] = frame.origin.x + frame.size.width
        result["bottom"] = frame.origin.y + frame.size.height
        return result
    }

    private func createLandmarkResult(_ landmark: FaceLandmark) -> JSObject {
        let positionResult = createPositionResult(landmark.position)
        var result = JSObject()
        result["type"] = landmark.type.rawValue
        result["position"] = positionResult
        return result
    }

    private func createPositionResult(_ point: VisionPoint) -> JSObject {
        var result = JSObject()
        result["x"] = Float(point.x)
        result["y"] = Float(point.y)
        return result
    }

    private func createContourResult(_ contour: FaceContour, _ pointsResult: JSArray) -> JSObject {
        var result = JSObject()
        result["type"] = contour.type.rawValue
        result["points"] = pointsResult
        return result
    }

    private func createPointsResult(_ points: [VisionPoint]) -> JSArray {
        var result = JSArray()

        for point in points {
            let positionResult = createPositionResult(point)
            result.append(positionResult)
        }

        return result
    }
}
